core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/jdbc/JDBCUserStoreManager.java

```diff
@@ -1362,6 +1637,133 @@ public boolean doAuthenticate(String userName, Object credential) throws UserSto
+                log.debug(sqlstmt);
+            }
+
+            prepStmt = dbConnection.prepareStatement(sqlstmt);
+            prepStmt.setString(1, preferredUserNameProperty);
+            prepStmt.setString(2, preferredUserNameValue);
+            prepStmt.setString(3, profileName);
+            if (sqlstmt.contains(UserCoreConstants.UM_TENANT_COLUMN)) {
+                prepStmt.setInt(4, tenantId);
+                prepStmt.setInt(5, tenantId);
+            }
+
+            rs = prepStmt.executeQuery();
+
+            int count = 0;
+            while (rs.next()) {
+
+                // Handle multiple matching users.
+                count++;
+                if (count > 1) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Invalid scenario. Multiple users found for the given username property: "
+                                + preferredUserNameProperty + " and value: " + preferredUserNameValue);
+                    }
+                    isAuthed = false;
+                    user = null;
+                    break;
+                }
+
+                String userID = rs.getString(1);
+                String storedPassword = rs.getString(2);
+                String saltValue = null;
+                if ("true".equalsIgnoreCase(
+                        realmConfig.getUserStoreProperty(JDBCRealmConstants.STORE_SALTED_PASSWORDS))) {
+                    saltValue = rs.getString(3);
+                }
+
+                boolean requireChange = rs.getBoolean(4);
+                Timestamp changedTime = rs.getTimestamp(5);
+
+                GregorianCalendar gc = new GregorianCalendar();
+                gc.add(GregorianCalendar.HOUR, -24);
+                Date date = gc.getTime();
+
+                if (requireChange && changedTime.before(date)) {
+                    isAuthed = false;
+                } else {
+                    password = this.preparePassword(credential, saltValue);
+                    if ((storedPassword != null) && (storedPassword.equals(password))) {
+                        isAuthed = true;
+                        RealmService realmService = UserCoreUtil.getRealmService();
+                        user = new User(userID,
+                                getUserClaimValueWithID(userID, UserCoreClaimConstants.USERNAME_CLAIM_URI, profileName),
+                                preferredUserNameValue, null, null, null);
+                        try {
+                            user.setTenantDomain(realmService.getTenantManager().getDomain(tenantId));
+                            user.setUserStoreDomain(UserCoreUtil.getDomainName(
+                                    CarbonContext.getThreadLocalCarbonContext().getUserRealm()
+                                            .getRealmConfiguration()));
+                        } catch (org.wso2.carbon.user.api.UserStoreException e) {
+                            throw new UserStoreException(e);
```

Comment:
Either throw the new error with more information or don't cat this error. Do the same thing in all the similar places.